ekncontent/ekncontent/eknc-database-manager.c

```diff
@@ -951,15 +910,50 @@ eknc_database_manager_query_internal (EkncDatabaseManager *self,
         xapian_enquire_set_cutoff (enquire, CLAMP (g_ascii_strtod (cutoff_str, NULL), 0, G_MAXUINT));
     }
 
+  unsigned limit, offset;
+  const char *offset_str = g_hash_table_lookup (query_options, QUERY_PARAM_OFFSET);
+  if (offset_str == NULL)
+    {
+      g_set_error_literal (error_out, EKNC_DATABASE_MANAGER_ERROR,
+                           EKNC_DATABASE_MANAGER_ERROR_INVALID_PARAMS,
+                           "Offset parameter is required for the query");
+      return NULL;
+    }
+
+  offset = CLAMP (g_ascii_strtod (offset_str, NULL), 0, G_MAXUINT);
+
+  const char *limit_str = g_hash_table_lookup (query_options, QUERY_PARAM_LIMIT);
+  if (limit_str == NULL)
+    {
+      g_set_error_literal (error_out, EKNC_DATABASE_MANAGER_ERROR,
+                           EKNC_DATABASE_MANAGER_ERROR_INVALID_PARAMS,
+                           "Limit parameter is required for the query");
+      return NULL;
+    }
+
+  /* str may contain a negative value to mean "all matching results"; since
+   * casting a negative floating point value into an unsigned integer is
+   * undefined behavior, we need to perform some level of validation first
+   */
+  {
```

Comment:
It's not really necessary to use a block, here: we're C99-only, we can place declarations after statements.